### PR TITLE
Jetpack Settings: Add module list fetching actions, reducer and selectors

### PR DESCRIPTION
**Pull request by oskosk**
Part of #9171. 

This PR introduces state handling for fetching a list of jetpack modules available in the self-hosted site via its API

#### Changes proposed in this Pull Request:

* Adds actions, reducer and selectors for fetching a site's list of modules.

#### Testing instructions

1. Run 
    ```
    $ npm run test-client client/state/jetpack-settings/
    ```
1. Expect the tests to pass.

### CODE REVIEW

File: client/state/action-types.js [hunk 1]
```diff
@@ -175,6 +175,9 @@ export const JETPACK_MODULE_ACTIVATE_FAILURE = 'JETPACK_MODULE_ACTIVATE_FAILURE'
 export const JETPACK_MODULE_DEACTIVATE = 'JETPACK_MODULE_DEACTIVATE';
 export const JETPACK_MODULE_DEACTIVATE_SUCCESS = 'JETPACK_MODULE_DEACTIVATE_SUCCESS';
 export const JETPACK_MODULE_DEACTIVATE_FAILURE = 'JETPACK_MODULE_DEACTIVATE_FAILURE';
+export const JETPACK_MODULE_LIST_REQUEST = 'JETPACK_MODULE_LIST_REQUEST';
+export const JETPACK_MODULE_LIST_REQUEST_SUCCESS = 'JETPACK_MODULE_LIST_REQUEST_SUCCESS';
+export const JETPACK_MODULE_LIST_REQUEST_FAILURE = 'JETPACK_MODULE_LIST_REQUEST_FAILURE';
```
Comment (tyxla) [inline client/state/action-types.js]: This is your call, of course, but if you ask me, I'd probably name those:

```
export const JETPACK_MODULES_RECEIVE = 'JETPACK_MODULES_RECEIVE';
export const JETPACK_MODULES_REQUEST = 'JETPACK_MODULES_REQUEST';
export const JETPACK_MODULES_REQUEST_FAILURE = 'JETPACK_MODULES_REQUEST_FAILURE';
export const JETPACK_MODULES_REQUEST_SUCCESS = 'JETPACK_MODULES_REQUEST_SUCCESS';
```

Notice that there are 4 action types, not 3. The idea here is to differentiate the action that indicates request success/failure from the action that receives the data (in this case, the module data). If you have a look at the rest of the actions that end in `_RECEIVE`, you'll see that's being done this way consistently throughout Calypso. See my line comment in actions.js for a demo of how I envision these 4 action types to work.

Comment (tyxla) [inline client/state/action-types.js]: This is your call, of course, but if you ask me, I'd probably name those:

```
export const JETPACK_MODULES_RECEIVE = 'JETPACK_MODULES_RECEIVE';
export const JETPACK_MODULES_REQUEST = 'JETPACK_MODULES_REQUEST';
export const JETPACK_MODULES_REQUEST_FAILURE = 'JETPACK_MODULES_REQUEST_FAILURE';
export const JETPACK_MODULES_REQUEST_SUCCESS = 'JETPACK_MODULES_REQUEST_SUCCESS';
```

Notice that there are 4 action types, not 3. The idea here is to differentiate the action that indicates request success/failure from the action that receives the data (in this case, the module data). If you have a look at the rest of the actions that end in `_RECEIVE`, you'll see that's being done this way consistently throughout Calypso. See my line comment in actions.js for a demo of how I envision these 4 action types to work.

File: client/state/jetpack-settings/modules/actions.js [hunk 2]
```diff
@@ -62,3 +70,28 @@ export const deactivateModule = ( siteId, moduleSlug ) => {
 			} );
 	};
 };
+
+export const fetchModuleList = ( siteId ) => {
```
Comment (tyxla) [inline client/state/jetpack-settings/modules/actions.js]: This looks great, I'm just suggesting the implementation with the 4 actions I've mentioned above.

```
export const fetchModuleList = ( siteId ) => {
	return ( dispatch ) => {
		dispatch( {
			type: JETPACK_MODULES_REQUEST,
			siteId
		} );

		return wp.undocumented().jetpackModules( siteId )
			.then( ( data ) => {
				const modules = keyBy( data.modules, 'id' );

				dispatch( {
					type: JETPACK_MODULES_RECEIVE,
					siteId,
					modules
				} );

				dispatch( {
					type: JETPACK_MODULES_REQUEST_SUCCESS,
					siteId
				} );
			} ).catch( error => {
				dispatch( {
					type: JETPACK_MODULES_REQUEST_FAILURE,
					siteId,
					error: error.message
				} );
			} );
	};
};
```

File: client/state/jetpack-settings/modules/actions.js [hunk 3]
```diff
@@ -1,4 +1,9 @@
 /**
+ * External dependencies
+ */
+import keyBy from 'lodash/keyBy';
```
Comment (tyxla) [inline client/state/jetpack-settings/modules/actions.js]: We're using direct imports from Lodash, Babel takes care of them:

```
import { keyBy } from 'lodash';
```

File: client/state/jetpack-settings/modules/test/actions.js [hunk 4]
```diff
@@ -3,13 +3,15 @@
  */
 import { expect } from 'chai';
 import sinon from 'sinon';
+import keyBy from 'lodash/keyBy';
```
Comment (tyxla) [inline client/state/jetpack-settings/modules/test/actions.js]: We're using direct imports from Lodash, Babel takes care of them:

```
import { keyBy } from 'lodash';
```

File: client/state/jetpack-settings/modules/reducer.js [hunk 5]
```diff
@@ -39,6 +50,14 @@ const createRequestsReducer = ( data ) => {
 	};
 };
 
+const createModuleListRequestReducer = ( data ) => {
```
Comment (tyxla) [inline client/state/jetpack-settings/modules/reducer.js]: If we're not planning to use this helper for anything else, we could inject `isFetchingModuleList` in there and accept just its boolean value as an argument:

```
const createModuleListRequestReducer = ( isFetchingModuleList ) => {
	return ( state, { siteId } ) => {
		return merge( {}, state, {
			[ siteId ]: {
				isFetchingModuleList
			}
		} );
	};
};
```

and then change the use in the  `requests` reducer to be shorter:

```
[ JETPACK_MODULE_LIST_REQUEST ]: createModuleListRequestReducer( true ),
[ JETPACK_MODULE_LIST_REQUEST_FAILURE ]: createModuleListRequestReducer( false ),
[ JETPACK_MODULE_LIST_REQUEST_SUCCESS ]: createModuleListRequestReducer( false )
```

File: client/state/jetpack-settings/modules/selectors.js [hunk 6]
```diff
@@ -37,3 +37,15 @@ export function isActivatingModule( state, siteId, moduleSlug ) {
...
+/**
+ * Returns true if we are currently making a request to get the list of Jetpack
+ * modules on the site
+ *
+ * @param  {Object}  state  Global state tree
+ * @param  {String}  siteId The ID of the site we're querying
+ * @return {?Boolean}         Whether the list is being requested
+ */
+export function isFetchingModuleList( state, siteId ) {
+	return get( state.jetpackSettings.jetpackModules.requests, [ siteId, 'isFetchingModuleList' ], null );
```
Comment (tyxla) [inline client/state/jetpack-settings/modules/selectors.js]: One thing we might want to think about here is consistency of the request property names. For the module activation/deactivation, they're called `activating` and `deactivating`, while here the prop is `isFetchingModuleList`. We might want to stay consistent, and either always prefix them with `is`, or never prefix them.

File: client/state/jetpack-settings/modules/test/selectors.js [hunk 7]
```diff
@@ -137,4 +138,19 @@ describe( 'selectors', () => {
...
+					jetpackSettings: {
+						jetpackModules: {
+							requests: REQUESTS_FIXTURE
+						}
+					}
+				},
+				siteId = 123456;
+			const output = isFetchingModuleList( stateIn, siteId );
+			expect( output ).to.be.true;
+		} );
```
Comment (tyxla) [inline client/state/jetpack-settings/modules/test/selectors.js]: Let's also add test cases for when the list of modules is not being fetched, and for when there is no such site at all (when it should return `null`).

File: client/state/jetpack-settings/modules/test/actions.js [hunk 8]
```diff
@@ -140,4 +149,61 @@ describe( 'actions', () => {
 			} );
 		} );
 	} );
+
+	describe( '#fetchModuleList', () => {
```
Comment (tyxla) [inline client/state/jetpack-settings/modules/test/actions.js]: If you decide to use the 4 action types, you'll have to update the tests accordingly.

**Comment by oskosk** [comment]
I've addressed a few of your observations @tyxla

- [x] Renaming JETPACK_MODULES_LIST_REQUEST* to JETPACK_MODULES_REQUEST*
- [x] Introducing JETPACK_MODULES_RECEIVE
- [x] Importing lodash keyBy method by destructuring from `lodash`.
- [x] Update action to dispatch JETPACK_MODULES_RECEIVE and JETPACK_MODULES_REQUEST_SUCCESS
- [x] Simplifying the parameter to createModuleListRequestReducer replacing object for a boolean,
- [x] Using consistent naming by renaming isFetchingModuleList to fetchingModules
- [ ] Adding test cases for when the list of modules is not being fetched, 
- [ ] Adding test case and for when there is no such site at all (when it should return null)

File: client/state/jetpack-settings/modules/actions.js [hunk 9]
```diff
@@ -62,3 +71,44 @@ export const deactivateModule = ( siteId, moduleSlug ) => {
...
+		dispatch( {
+			type: JETPACK_MODULES_REQUEST,
+			siteId
+		} );
+
+		return wp.undocumented().jetpackModules( siteId )
+			.then( ( data ) => {
+				const modules = keyBy( data.modules, 'id' );
+				dispatch( receiveJetpackModules( siteId, modules ) );
+				dispatch( {
```
Comment (johnHackworth) [inline client/state/jetpack-settings/modules/actions.js]: why dispatching two different events here? (JEPTACK_MODULES_RECEIVE and JEPTACK_MODULES_REQUEST_SUCCESS) . Can we use only one and subscript to it whatever reducer uses the other?

File: client/state/jetpack-settings/modules/selectors.js [hunk 10]
```diff
@@ -37,3 +37,15 @@ export function isActivatingModule( state, siteId, moduleSlug ) {
...
+/**
+ * Returns true if we are currently making a request to get the list of Jetpack
+ * modules on the site
+ *
+ * @param  {Object}  state  Global state tree
+ * @param  {String}  siteId The ID of the site we're querying
+ * @return {?Boolean}         Whether the list is being requested
+ */
+export function isFetchingModules( state, siteId ) {
+	return get( state.jetpackSettings.jetpackModules.requests, [ siteId, 'fetchingModules' ], null );
```
Comment (johnHackworth) [inline client/state/jetpack-settings/modules/selectors.js]: I think the default here should be `false`, not `null`. Being a "isWhatever" method you expect a boolean response

**Comment by johnHackworth** [comment]
I left a minor comment, but apart from that, I think this is ready to go :ship:

**Pull request merged**